core/base/v7/inc/ROOT/TDirectory.h

```diff
@@ -148,35 +148,31 @@ class TDirectory {
   /// \throws TDirectoryTypeMismatch if the object stored under this name is of
   ///   a type that is not a derived type of `T`.
   template <class T>
-  std::shared_ptr<T> Get(const std::string& name) {
-    ContentMap_t::iterator idx = fContent.find(name);
+  decltype(auto) Get(const std::string& name) {
+    auto idx = fContent.find(name);
     if (idx != fContent.end()) {
-      // FIXME: implement upcast!
-      if (auto dep
-        = dynamic_cast<Internal::TDirectoryEntryPtr<T>*>(idx->second.get())) {
-        return dep->GetPointer();
-      }
+      if (idx->second.GetTypeInfo() == typeid(ToContentType_t<T>))
```

Comment:
Sorry, I meant to write `type_info` not `unique_ptr`!

And you're right, same for MSVC: http://stackoverflow.com/questions/27121932/can-anyone-explain-this-out-of-bounds-array-access-in-the-msvc-std-library

So your code is fine as is; ignore my comment, please.